lib/charms/tls_certificates_interface/v2/tls_certificates.py

```diff
@@ -1159,6 +1168,84 @@ def _revoke_certificates_for_which_no_csr_exists(self, relation_id: int) -> None
                 )
                 self.remove_certificate(certificate=certificate["certificate"])
 
+    def get_requirer_units_csrs_with_no_certs(
+        self,
+    ) -> List[Dict[str, Union[int, str, List[Dict[str, str]]]]]:
+        """Filters the requirer's units csrs.
+
+        Keeps the ones for which no certificate was provided.
+
+        Returns:
+            list: List of dictonries that contain the unit's csrs
+            that don't have a certificate issued.
+        """
+        all_unit_csr_mappings = copy.deepcopy(self.get_requirer_csrs_by_unit())
+        for unit_csrs in all_unit_csr_mappings:
+            base64_keys_list = list(
+                map(
+                    lambda item: base64.b64encode(item.encode()).decode(),
+                    self.get_issued_certificates(unit_csrs["relation_id"])[  # type: ignore[arg-type]
+                        unit_csrs["application_name"]  # type: ignore[index]
+                    ].keys(),
+                )
+            )
```

Comment:
This is hard to understand. Please prioritise readability/simplicity over efficiency. We may want a method to check whether a cert was issued for a given CSR.

Reply:
I agree, I'm not even 100% confident this should be done by the library.
I moved it to the operator, at the end it is the action that should output the csr in base64, while we don't use the csr in base64 somewhere else in the library.

I added a function that validates if a certificate was issued.